openslides/utils/collection.py

```diff
@@ -117,14 +152,217 @@ def get_full_data(self):
+
+        # Add the id of the element to the collection
+        Collection(self.collection_string).add_id_to_cache(self.id)
+
+
+class Collection:
+    """
+    Represents all elements of one collection.
+    """
+
+    def __init__(self, collection_string):
+        self.collection_string = collection_string
+
+    def get_cache_key(self):
+        """
+        Returns a string that is used as cache key for a collection.
+        """
+        return get_element_list_cache_key(self.collection_string)
+
+    def get_model(self):
+        """
+        Returns the django model that is used for this collection.
+        """
+        return get_model_from_collection_string(self.collection_string)
+
+    def element_generator(self):
+        """
+        Generator that yields all collection_elements of this collection.
+        """
+        # Get all cache keys.
+        ids = self.get_all_ids()
+        cache_keys = [
+            get_single_element_cache_key(self.collection_string, id)
+            for id in ids]
+        cached_full_data_dict = cache.get_many(cache_keys)
+
+        # Get all ids that are missing.
+        missing_ids = set()
+        for cache_key in cache_keys:
+            if cache_key not in cached_full_data_dict:
+                missing_ids.add(get_collection_id_from_cache_key(cache_key)[1])
+
+        # Generate collection elements that where is the cache.
+        for cache_key, cached_full_data in cached_full_data_dict.items():
+            yield CollectionElement.from_values(
+                *get_collection_id_from_cache_key(cache_key),
+                full_data=cached_full_data)
+
+        # Generate collection element that where not in the cache.
+        # If config elements are not in the cache, they have to be read from
+        # the cache object.
+        from openslides.core.config import config
+        if self.collection_string == config.get_collection_string():
+            for key in missing_ids:
+                yield CollectionElement.from_values(config.get_collection_string(), key)
+        else:
+            for instance in self.get_model().objects.filter(pk__in=missing_ids):
+                yield CollectionElement.from_instance(instance)
+
+    def as_autoupdate_for_projector(self):
+        """
+        Returns a list of dictonaries to send them to the projector.
+        """
+        output = []
+        for collection_element in self.element_generator():
+            content = collection_element.as_autoupdate_for_projector()
+            if content is not None:
+                output.append(content)
+        return output
+
+    def as_list_for_user(self, user):
+        """
+        Returns a list of dictonaries to send them to a user, for example over
+        the rest api.
+        """
+        output = []
+        for collection_element in self.element_generator():
+            content = collection_element.as_dict_for_user(user)
+            if content is not None:
```

Comment:
No. Here it is different. as_dict_for_user returns access_permissions.restricted_data(user) which can return None.